apps/prairielearn/src/schemas/schemas/infoAssessment.ts

```diff
@@ -0,0 +1,373 @@
+    uids: z
+      .array(z.string())
+      .describe("A list of UIDs (like 'username@example.com', one of which is required for access")
+      .optional(),
+    credit: z
+      .number()
+      .int()
+      .gte(0)
+      .describe(
+        'How much credit is awarded for doing the homework, as a percentage (100 means full credit).',
+      )
+      .optional(),
+    startDate: z.string().describe('The earliest date on which access is permitted.').optional(),
+    endDate: z.string().describe('The latest date on which access is permitted.').optional(),
+    timeLimitMin: z
+      .number()
+      .int()
+      .gte(0)
+      .describe('The time limit to complete the assessment, in minutes (only for Exams).')
+      .optional(),
+    password: z.string().describe('Password to begin the assessment (only for Exams).').optional(),
+    showClosedAssessment: z
+      .boolean()
+      .describe('Whether the student can view the assessment after it has been closed')
+      .optional(),
+    showClosedAssessmentScore: z
+      .boolean()
+      .describe(
+        'Whether the student can view the assessment grade after it has been closed. Only works if showClosedAssessment is also set to false',
+      )
+      .optional(),
+    active: z
+      .boolean()
+      .describe(
+        'Whether the student can create a new assessment instance and submit answers to questions. If set to false, the available credit must be 0.',
+      )
+      .optional(),
+  })
+  .strict()
+  .describe(
+    'An access rule that permits people to access this assessment. All restrictions in the rule must be satisfied for the rule to allow access.',
+  );
+
+export type AsssessmentAccessRule = z.infer<typeof AsssessmentAccessRuleSchema>;
+
+export const PointsSingleSchema = z.number().gte(0).describe('A single point value.');
+
+export const PointsListSchema = z
+  .array(PointsSingleSchema)
+  .min(1)
+  .describe('An array of point values.');
+
+export const PointsSchema = z.union([PointsSingleSchema, PointsListSchema]);
+
+export const QuestionIdSchema = z
+  .string()
+  .describe('Question ID (directory name of the question).');
+
+export const ForceMaxPointsSchema = z
+  .boolean()
+  .describe('Whether to force this question to be awarded maximum points on a regrade.');
+
+export const AdvanceScorePercSchema = z
+  .number()
+  .gte(0)
+  .lte(100)
+  .describe('Minimum score percentage to unlock access to subsequent questions');
+
+const QuestionAlternativeSchema = z.object({
+  comment: z
+    .union([z.string(), z.array(z.any()), z.object({}).catchall(z.any())])
+    .describe('Arbitrary comment for reference purposes.')
+    .optional(),
+  points: PointsSchema.optional(),
+  autoPoints: PointsSchema.optional(),
+  maxPoints: PointsSingleSchema.optional(),
+  maxAutoPoints: PointsSingleSchema.optional(),
+  manualPoints: PointsSingleSchema.optional(),
+  id: QuestionIdSchema, // .optional(),
```

Comment:
The code here: https://github.com/PrairieLearn/PrairieLearn/blob/33d10056121c78138c735122c031fbe5a737688a/apps/prairielearn/src/sync/course-db.ts#L1178 requires this to exist.

Reply:
This one I am less sure on then the others -- surely an undefined alternative id would have triggered a bug by now in checkAndRecordQid if it was optional?

Reply:
The check that checks for '@' was introduced two years after the original code was written, by Seth.

Reply:
This line:
```
const questionId = questionIds[alternative.qid];
```
leads me to believe that it should be required and is not currently required.

Reply:
Resolving; made required in the schema.